src/test/java/htsjdk/samtools/CRAMComplianceTest.java

```diff
@@ -147,4 +175,156 @@ private void assertSameRecords(int majorVersion, SAMRecord record1, SAMRecord re
+        writeRecordsToFile(bamRecords, tempCRAMFile, referenceFile, samHeader);
+        List<SAMRecord> roundTripCRAMRecords = getSAMRecordsFromFile(tempCRAMFile, referenceFile);
+
+        // compare to originals
+        i = 0;
+        for (SAMRecord rec : roundTripCRAMRecords) {
+            Assert.assertTrue(rec.equals(originalCRAMRecords.get(i++)));
+        }
+        Assert.assertEquals(i, originalCRAMRecords.size());
+    }
+
+    @Test
+    public void testBAMThroughCRAMRoundTrip() throws IOException, NoSuchAlgorithmException {
+        final File TEST_DATA_DIR = new File("src/test/resources/htsjdk/samtools/cram");
+
+        // These files are reduced versions of the CEUTrio.HiSeq.WGS.b37.NA12878.20.21.bam and human_g1k_v37.20.21.fasta
+        // files used in GATK4 tests. The first 8000 records from chr20 were extracted; from those around 80 placed but
+        // unmapped reads that contained cigar elements were removed, along with one read who's mate was on chr21.
+        // Finally all read positions were remapped to the subsetted reference file, which contains only the ~9000 bases
+        // used by the reduced read set.
+        final File originalBAMInputFile = new File(TEST_DATA_DIR, "CEUTrio.HiSeq.WGS.b37.NA12878.20.first.8000.bam");
+        final File referenceFile = new File(TEST_DATA_DIR, "human_g1k_v37.20.subset.fasta");
+
+        // retrieve all records from the bam and reset the indexing bins to keep comparisons with
+        // cram records from failing
+        List<SAMRecord> originalBAMRecords = getSAMRecordsFromFile(originalBAMInputFile, referenceFile);
+        for (int i = 0; i < originalBAMRecords.size(); i++) {
+            originalBAMRecords.get(i).setIndexingBin(null);
+        }
+
+        // write the BAM records to a temporary CRAM
+        final File tempCRAMFile = File.createTempFile("testBAMThroughCRAMRoundTrip", CramIO.CRAM_FILE_EXTENSION);
+        tempCRAMFile.deleteOnExit();
+        SAMFileHeader samHeader = getFileHeader(originalBAMInputFile, referenceFile);
+        writeRecordsToFile(originalBAMRecords, tempCRAMFile, referenceFile, samHeader);
+
+        // read the CRAM records back in and compare to the original BAM records
+        List<SAMRecord> cramRecords = getSAMRecordsFromFile(tempCRAMFile, referenceFile);
+        Assert.assertEquals(cramRecords.size(), originalBAMRecords.size());
+        for (int i = 0; i < originalBAMRecords.size(); i++) {
+            Assert.assertEquals(originalBAMRecords.get(i), cramRecords.get(i));
+        }
+
+        //  write the roundtripped CRAM records out to a bam and compare to the original
+        final File roundTrippedBAMFile = File.createTempFile("testBAMThroughCRAMRoundTrip", BamFileIoUtils.BAM_FILE_EXTENSION);
+        roundTrippedBAMFile.deleteOnExit();
+        writeRecordsToFile(cramRecords, roundTrippedBAMFile, referenceFile, samHeader);
+
+        Assert.assertTrue(compareMD5Digests(roundTrippedBAMFile, originalBAMInputFile));
```

Comment:
not sure about binary equality of BAM files, for example changing compression level will break this.